tempesta_fw/cache.c

```diff
@@ -1874,13 +1910,27 @@ tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p)
 	}
 	BUG_ON(it->frag < 0);
 
+	frame_hdr.type = HTTP2_DATA;
+
 	while (1) {
 		int off, f_size;
 
 		/* TDB keeps data by pages and we can reuse the pages. */
 		off = (unsigned long)p & ~PAGE_MASK;
 		f_size = trec->data + trec->len - p;
 		if (f_size) {
+			if (h2) {
+				frame_hdr.length = f_size;
+				body_sz -= f_size;
+				frame_hdr.flags = body_sz ? 0
+							  : HTTP2_F_END_STREAM;
+				tfw_h2_pack_frame_header(buf, &frame_hdr);
+
+				r = __append_h2_frame_header(it, &frame_hdr_str);
```

Comment:
> We discussed that HTTP/2 is the first class citizen in Tempesta, so why didn't you just reserve the 9 bytes for each page in the TDB?

Each response generated from cache will have unique `stream_id` field in the frame header, so we can't just write to the same page from multiple responses, data race will happen. 

>  That's not good at all.
Agreed. And you got the idea right. I should use a page(s) used to store http headers, since frame headers are really small, It shouldn't be a problem to place data there. And I should reduce amount of fragments with frame headers: since all clients must support at least 16kb frames, the number of fragments with headers will be 4 times smaller than now.

Reply:
> Each response generated from cache will have unique stream_id field in the frame header, so we can't just write to the same page from multiple responses, data race will happen.

Yeah... We copy the pages from the cache anyway for HTTP/2 transmission, so may be we can copy the data right in `cache.c` code with extension for the frament headers  instead of setting `SKBTX_SHARED_FRAG` flag and do the copy on TCP transmission? Just consider the option.